provider/oci/environ.go

```diff
@@ -223,42 +378,313 @@ func (e *Environ) Destroy(ctx envcontext.ProviderCallContext) error {
+		args.InstanceConfig,
+		cloudcfg,
+		OCIRenderer{},
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot make user data")
+	}
+
+	var rootDiskSize int
+	if args.Constraints.RootDisk != nil {
+		rootDiskSize = int(*args.Constraints.RootDisk)
+		if int(*args.Constraints.RootDisk) < MinVolumeSizeMB {
+			rootDiskSize = MinVolumeSizeMB / 1024
+		} else if int(*args.Constraints.RootDisk) > MinVolumeSizeMB {
+			rootDiskSize = MaxVolumeSizeMB / 1024
+		}
+	} else {
+		rootDiskSize = MinVolumeSizeMB / 1024
+	}
+
+	// NOTE(gsamfira): Should we make this configurable?
+	// TODO(gsamfira): select Availability Domain and subnet ID
+	assignPublicIp := true
+	bootSource := ociCore.InstanceSourceViaImageDetails{
+		ImageId:             &image,
+		BootVolumeSizeInGBs: &rootDiskSize,
+	}
+	instanceDetails := ociCore.LaunchInstanceDetails{
+		AvailabilityDomain: &zone,
+		CompartmentId:      e.ecfg().compartmentID(),
+		SourceDetails:      bootSource,
+		Shape:              &spec.InstanceType.Name,
+		CreateVnicDetails: &ociCore.CreateVnicDetails{
+			SubnetId:       network.Id,
+			AssignPublicIp: &assignPublicIp,
+			DisplayName:    &hostname,
+		},
+		DisplayName: &hostname,
+		Metadata: map[string]string{
+			"user_data": string(userData),
+		},
+		FreeformTags: tags,
+	}
+
+	request := ociCore.LaunchInstanceRequest{
+		LaunchInstanceDetails: instanceDetails,
+	}
+
+	response, err := e.Compute.LaunchInstance(context.Background(), request)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	instance, err := newInstance(response.Instance, e)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	machineId := response.Instance.Id
+	timeout := 10 * time.Minute
+	if err := instance.waitForMachineStatus(desiredStatus, timeout); err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Infof("started instance %q", *machineId)
+
+	if desiredStatus == ociCore.InstanceLifecycleStateRunning {
+		if err := instance.waitForPublicIP(ctx); err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	result := &environs.StartInstanceResult{
+		Instance: instance,
+		Hardware: instance.hardwareCharacteristics(),
+	}
+
+	return result, nil
 }
 
 // StopInstances implements environs.InstanceBroker.
 func (e *Environ) StopInstances(ctx envcontext.ProviderCallContext, ids ...instance.Id) error {
-	return errors.NotImplementedf("StopInstances")
+	ociInstances, err := e.getOciInstances(ids...)
+	if err == environs.ErrNoInstances {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	logger.Debugf("terminating instances %v", ids)
+	if err := e.terminateInstances(ociInstances...); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *Environ) terminateInstances(instances ...*ociInstance) error {
+	wg := sync.WaitGroup{}
+	wg.Add(len(instances))
+	errs := []error{}
+	instIds := []instance.Id{}
+	for _, oInst := range instances {
+		go func(inst *ociInstance) {
+			defer wg.Done()
+			if err := inst.deleteInstance(); err != nil {
```

Comment:
Do we know that the Oracle API library properly handles concurrent requests? Hopefully you tested this and it works fine. I certainly do like that if we are waiting for things to die that we can wait in parallel for 5 minutes, we wouldn't want to wait sequentially.

Reply:
Considering it's requests for different resources to be deleted, it should not be a problem. As you can imagine, I have not tested with more than 5 machines running 1 unit each, as I did not have access to more resources then that :). I imagine that doing 2000 concurrent requests will probably fail. I can batch them in smaller groups, to make sure we don't get throttled. How does that sound?

Reply:
I was concerned just from a goroutine "could you end up with 2 things writing to the same socket causing invalid requests".
I see your point about throttling, but I'm happy to get-there-when-we-need it.

Reply:
Each call to ```TerminateInstance()``` will do a separate http call. If the go http package fails in such a way, we should probably revise all code that uses it :).